File: Day 5/Day 5.py
def parse_input(filename: str):
    with open(filename, "r") as file:
        lines = []
        for line in file:
            lines.append(line.strip())
        return lines


def get_offset(letter: str, total_seats: int):
    offset = -1
    if letter == "F" or letter == "L":
        offset = 0
    elif letter == "B" or letter == "R":
        offset = total_seats // 2

    return offset


def interpret_pass(boarding_pass: str):
    current_num = 128
    offsets = []
    for i in range(7):
        offsets.append(get_offset(boarding_pass[i], current_num))
        current_num //= 2
    row = sum(offsets)
    offsets = []
    current_num = 8
    for i in range(7, 10):
        offsets.append(get_offset(boarding_pass[i], current_num))
        current_num //=2
    column = sum(offsets)
    return row, column


def main():
    passes = parse_input("input.txt")
    seats = []
    for boarding_pass in passes:
        row, column = interpret_pass(boarding_pass)
        seats.append(row * 8 + column)
    max_seat_ID = max(seats)
    min_seat_ID = min(seats)
    print(f"The highest seat ID is {max_seat_ID}")
    print(f"The lowest seat ID is {min_seat_ID}")

    ids_to_check = [x for x in range(min_seat_ID + 1, max_seat_ID)]  # don't need to check max and min
    for seat_id in ids_to_check:
        if seat_id not in seats:
            print(f"The seat ID {seat_id} is not in the list of seats")
            break  # we've been told there should only be one, so we can stop when we find it


if __name__ == "__main__":
    main()

File: Day 5/Day 5.cpp
#include <vector>
#include <string>
#include <fstream>
#include <numeric>
#include <algorithm>
#include <iostream>

std::vector<std::string> parse_input(const std::string& filename) {
	std::ifstream file(filename);
	std::string line;
	std::vector<std::string> output;
	while (std::getline(file, line)) {
		output.emplace_back(line);
	}
	return output;
}

int get_offset(char letter, int total_seats) {
	int offset = -1;
	if (letter == 'F' || letter == 'L')
		offset = 0;
	else if (letter == 'B' || letter == 'R')
		offset = total_seats / 2;

	return offset;
}

std::pair<int, int> interpret_pass(const std::string& boarding_pass) {
	int current_num = 128;
	std::vector<int> offsets;
	for (int i = 0; i < 7; i++) {
		offsets.push_back(get_offset(boarding_pass[i], current_num));
		current_num /= 2;
	}
	int row = std::accumulate(offsets.begin(), offsets.end(), 0);
	current_num = 8;
	offsets.clear();
	for (int i = 7; i < 10; i++) {
		offsets.push_back(get_offset(boarding_pass[i], current_num));
		current_num /= 2;
	}
	int column = std::accumulate(offsets.begin(), offsets.end(), 0);
	return std::pair<int, int>(row, column);
}

int main() {
	auto passes = parse_input("input.txt");
	std::vector<int> seats;
	for (const auto& boarding_pass : passes) {
		// 1st = row, 2nd = column
		std::pair<int, int> seat_pos = interpret_pass(boarding_pass);
		seats.push_back(seat_pos.first * 8 + seat_pos.second);
	}
	std::sort(seats.begin(), seats.end());
	// Because we just sorted, we know where these are; no need to search for them.
	int max_seat_ID = *(seats.end() - 1);
	int min_seat_ID = *seats.begin();
	std::cout << "The highest seat ID is " << max_seat_ID << "\n";
	std::cout << "The lowest seat ID is " << min_seat_ID << "\n";

	// Assumes seats is sorted (which it is)
	for (int i = 1; i < static_cast<int>(seats.size()); i++) {
		if (seats[i] != seats[i - 1] + 1) {
			std::cout << "Our seat ID is " << seats[i - 1] + 1 << "\n";
			break;
		}
	}

}